File: iOS/ssabree/ssabree/Features/MyPage/MyPageViewModel.swift
import Foundation

// MARK: - MyPage UI State

struct MyPageUiState {
    var isLoading: Bool = false
    var isLoadingPortfolio: Bool = false
    var isUploadingImage: Bool = false
    var errorMessage: String? = nil
    var myPage: MyPageModel? = nil
    var stackImageMap: [String: String] = [:]
}

// MARK: - MyPage ViewModel

@Observable
final class MyPageViewModel {
    private let myPageRepository: MyPageRepository
    private let portfolioRepository: PortfolioRepository?
    private let uploadRepository: UploadRepository?
    private let stackRepository: StackRepository?

    private(set) var uiState = MyPageUiState()

    init(myPageRepository: MyPageRepository, portfolioRepository: PortfolioRepository? = nil, uploadRepository: UploadRepository? = nil, stackRepository: StackRepository? = nil) {
        self.myPageRepository = myPageRepository
        self.portfolioRepository = portfolioRepository
        self.uploadRepository = uploadRepository
        self.stackRepository = stackRepository
    }

    @MainActor
    func loadMyPage() async {
        let hasCached = uiState.myPage != nil
        uiState.isLoading = !hasCached
        uiState.errorMessage = nil

        async let myPageResult = myPageRepository.getMyPage()
        async let stacksResult = loadStackImageMap()
        async let portfoliosResult = loadPortfolios()

        let stackMap = await stacksResult
        let portfolios = await portfoliosResult
        let result = await myPageResult

        switch result {
        case .success(var myPage):
            print("[MyPageViewModel] loadMyPage: succeed")
            // 포트폴리오에서 solved.ac 정보 보강 (백엔드 MyPage API는 solvedAcRank만 반환)
            if let portfolio = portfolios.first {
                myPage = enrichWithPortfolio(myPage: myPage, portfolio: portfolio, allPortfolios: portfolios)
            }
            uiState.myPage = myPage
            uiState.stackImageMap = stackMap
            uiState.isLoading = false
        case .failure(let error):
            print("[MyPageViewModel] loadMyPage: failed (\(error.localizedDescription))")
            uiState.errorMessage = error.localizedDescription
            uiState.isLoading = false
        }
    }

    private func loadPortfolios() async -> [PortfolioModel] {
        guard let portfolioRepository = portfolioRepository else { return [] }
        let result = await portfolioRepository.getMyPortfolios()
        switch result {
        case .success(let portfolios): return portfolios
        case .failure: return []
        }
    }

    private func enrichWithPortfolio(myPage: MyPageModel, portfolio: PortfolioModel, allPortfolios: [PortfolioModel]) -> MyPageModel {
        let existing = myPage.portfolioSummary
        let solvedAcInfo = portfolio.solvedAcInfo

        let enrichedSummary = MyPagePortfolioSummaryModel(
            techStack: existing?.techStack ?? [:],
            ssafySwRating: existing?.ssafySwRating ?? portfolio.swTestRank,
            solvedAcRank: existing?.solvedAcRank,
            solvedAcHandle: portfolio.bojHandle,
            solvedAcTierName: solvedAcInfo?.tierName,
            solvedAcTierImageUrl: solvedAcInfo?.tierImageUrl,
            solvedAcSolvedCount: solvedAcInfo?.solvedCount,
            links: existing?.links ?? portfolio.urls.map { $0.url },
            projects: existing?.projects ?? allPortfolios.map { $0.title }
        )

        return MyPageModel(
            user: myPage.user,
            counts: myPage.counts,
            portfolioSummary: enrichedSummary
        )
    }

    private func loadStackImageMap() async -> [String: String] {
        guard let stackRepository = stackRepository else { return [:] }
        let result = await stackRepository.getStacks()
        switch result {
        case .success(let stacks):
            var map: [String: String] = [:]
            for stack in stacks {
                if let imgUrl = stack.imgUrl, !imgUrl.isEmpty {
                    map[stack.name] = imgUrl
                }
            }
            return map
        case .failure:
            return [:]
        }
    }

    @MainActor
    func uploadProfileImage(imageData: Data) async {
        guard let uploadRepository = uploadRepository else { return }

        uiState.isUploadingImage = true
        uiState.errorMessage = nil

        let uploadResult = await uploadRepository.uploadImage(image: imageData)

        switch uploadResult {
        case .success(let imageUrl):
            // 프로필 이미지 URL 업데이트
            let updateResult = await myPageRepository.updateProfileImage(imageUrl)
            switch updateResult {
            case .success:
                print("[MyPageViewModel] uploadProfileImage: succeed")
                await loadMyPage()
            case .failure(let error):
                print("[MyPageViewModel] updateProfileImage: failed (\(error.localizedDescription))")
                uiState.errorMessage = error.localizedDescription
            }
        case .failure(let error):
            print("[MyPageViewModel] uploadProfileImage: failed (\(error.localizedDescription))")
            uiState.errorMessage = error.localizedDescription
        }

        uiState.isUploadingImage = false
    }

    @MainActor
    func deleteProfileImage() async {
        uiState.isUploadingImage = true

        let result = await myPageRepository.updateProfileImage("")

        switch result {
        case .success:
            print("[MyPageViewModel] deleteProfileImage: succeed")
            await loadMyPage()
        case .failure(let error):
            print("[MyPageViewModel] deleteProfileImage: failed (\(error.localizedDescription))")
            uiState.errorMessage = error.localizedDescription
        }

        uiState.isUploadingImage = false
    }

    @MainActor
    func setError(_ message: String) {
        uiState.errorMessage = message
    }
}

File: android/ssabree/app/src/main/java/com/ssafy/ssabree/features/mypage/screen/MyPageViewModel.kt
package com.ssafy.ssabree.features.mypage.screen

import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.ssafy.ssabree.app.ApplicationClass
import com.ssafy.ssabree.core.datasource.local.CachedMyPage
import com.ssafy.ssabree.core.datasource.local.MyPageLocalStore
import com.ssafy.ssabree.core.repository.MemberRepository
import com.ssafy.ssabree.core.repository.PortfolioRepository
import com.ssafy.ssabree.core.repository.ProjectRepository
import com.ssafy.ssabree.core.repository.UploadRepository
import com.ssafy.ssabree.core.repository.model.MyPageModel
import com.ssafy.ssabree.core.repository.model.MyPagePortfolioSummaryModel
import com.ssafy.ssabree.core.repository.model.PortfolioModel
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.io.File

private const val TAG = "MyPageViewModel"

data class MyPageUiState(
    val isLoading: Boolean = false,
    val isLoadingPortfolio: Boolean = false,
    val isUploadingImage: Boolean = false,
    val errorMessage: String? = null,
    val myPage: MyPageModel? = null
)

class MyPageViewModel(
    private val memberRepository: MemberRepository,
    private val portfolioRepository: PortfolioRepository,
    private val projectRepository: ProjectRepository,
    private val uploadRepository: UploadRepository? = null
) : ViewModel() {
    private val _uiState = MutableStateFlow(MyPageUiState())
    val uiState: StateFlow<MyPageUiState> = _uiState.asStateFlow()
    private val localStore = MyPageLocalStore(ApplicationClass.appContext)

    init {
        localStore.loadMyPage()?.myPage?.let { cached ->
            _uiState.update { it.copy(isLoading = false, myPage = cached) }
        }
        loadMyPage()
    }

    fun loadMyPage() = viewModelScope.launch {
        val hasCached = _uiState.value.myPage != null
        val hasPortfolioCached = _uiState.value.myPage?.portfolioSummary != null
        _uiState.update { it.copy(isLoading = !hasCached, isLoadingPortfolio = !hasPortfolioCached, errorMessage = null) }

        coroutineScope {
            val memberDeferred = async { memberRepository.getMyPage() }
            val portfolioDeferred = async { portfolioRepository.getMyPortfolios() }
            val scrapsDeferred = async { memberRepository.getMyScraps() }

            val memberResult = memberDeferred.await()
            val scrapsResult = scrapsDeferred.await()
            val scrapCountOverride = scrapsResult.getOrNull()?.size?.toLong()
            memberResult
                .onSuccess { myPage ->
                    Log.d(TAG, "loadMyPage: member succeed")
                    val adjustedCounts = myPage.counts?.let { counts ->
                        if (scrapCountOverride != null) {
                            counts.copy(scrapCount = scrapCountOverride)
                        } else {
                            counts
                        }
                    }
                    val adjusted = if (adjustedCounts != myPage.counts) {
                        myPage.copy(counts = adjustedCounts)
                    } else {
                        myPage
                    }
                    _uiState.update { it.copy(isLoading = false, myPage = adjusted) }
                    localStore.saveMyPage(CachedMyPage(adjusted))
                }
                .onFailure { e ->
                    Log.d(TAG, "loadMyPage: member failed (${e.message})")
                    _uiState.update { it.copy(isLoading = false, errorMessage = e.message) }
                }

            val portfolioResult = portfolioDeferred.await()
            val portfolio = portfolioResult.getOrNull()?.firstOrNull()
            val projectTitles = portfolio?.let { current ->
                projectRepository.getProjectsByPortfolio(current.id)
                    .getOrNull()
                    ?.map { project -> project.title }
                    .orEmpty()
            }.orEmpty()
            val summary = portfolio?.toSummary(projectTitles)

            _uiState.update { state ->
                if (summary != null) {
                    val base = state.myPage ?: MyPageModel(user = null, counts = null, portfolioSummary = null)
                    val merged = base.copy(portfolioSummary = summary)
                    localStore.saveMyPage(CachedMyPage(merged))
                    state.copy(isLoadingPortfolio = false, myPage = merged)
                } else {
                    state.copy(isLoadingPortfolio = false)
                }
            }
        }
    }

    fun uploadProfileImage(file: File) = viewModelScope.launch {
        val repo = uploadRepository ?: return@launch
        _uiState.update { it.copy(isUploadingImage = true) }
        repo.uploadImage(file)
            .onSuccess { imageUrl ->
                memberRepository.updateProfileImage(imageUrl)
                    .onSuccess {
                        Log.d(TAG, "uploadProfileImage: succeed")
                        loadMyPage()
                    }
                    .onFailure { e ->
                        Log.e(TAG, "updateProfileImage: failed (${e.message})")
                        _uiState.update { it.copy(isUploadingImage = false, errorMessage = e.message) }
                    }
            }
            .onFailure { e ->
                Log.e(TAG, "uploadProfileImage: failed (${e.message})")
                _uiState.update { it.copy(isUploadingImage = false, errorMessage = e.message) }
            }
        _uiState.update { it.copy(isUploadingImage = false) }
    }

    /**
     * 프로필 이미지 삭제 (빈 값으로 업데이트)
     */
    fun deleteProfileImage() = viewModelScope.launch {
        _uiState.update { it.copy(isUploadingImage = true) }
        memberRepository.updateProfileImage("")
            .onSuccess {
                Log.d(TAG, "deleteProfileImage: succeed")
                loadMyPage()
            }
            .onFailure { e ->
                Log.e(TAG, "deleteProfileImage: failed (${e.message})")
                _uiState.update { it.copy(isUploadingImage = false, errorMessage = e.message) }
            }
        _uiState.update { it.copy(isUploadingImage = false) }
    }
}

private fun PortfolioModel.toSummary(projectTitles: List<String>): MyPagePortfolioSummaryModel {
    val techStack = stacks.associate { stack ->
        stack.stackName to (stack.expertLevel ?: "")
    }
    val links = urls.map { url -> url.url }
    return MyPagePortfolioSummaryModel(
        techStack = techStack,
        ssafySwRating = swTestRank,
        solvedAcRank = solvedacRank,
        solvedAcHandle = bojHandle,
        solvedAcTierName = solvedAcInfo?.tierName,
        solvedAcTierImageUrl = solvedAcInfo?.tierImageUrl,
        solvedAcSolvedCount = solvedAcInfo?.solvedCount,
        links = links,
        projects = projectTitles
    )
}
